server/src/main/java/org/opensearch/search/aggregations/Aggregator.java

```diff
@@ -185,13 +195,15 @@ public interface BucketComparator {
 
     /**
      * Build the result of this aggregation if it is at the "top level"
-     * of the aggregation tree. If, instead, it is a sub-aggregation of
-     * another aggregation then the aggregation that contains it will call
-     * {@link #buildAggregations(long[])}.
+     * of the aggregation tree and save it. This should get called
+     * during post collection. If, instead, it is a sub-aggregation
+     * of another aggregation then the aggregation that contains
+     * it will call {@link #buildAggregations(long[])}.
      */
     public final InternalAggregation buildTopLevel() throws IOException {
         assert parent() == null;
-        return buildAggregations(new long[] { 0 })[0];
+        this.internalAggregation.set(buildAggregations(new long[] { 0 })[0]);
```

Comment:
is there any assumptions for profiler that buildAggregation is called sequentially in concurrent case. Probably good to double check  that.

Reply:
Added a ConcurrentHashMap in the latest revision. I did check that the timers themselves are thread safe.

Reply:
Took a closer look at this -- I think it should be fine. The `AggregationProfileBreakdown` object is created per Aggregator and set in a single-threaded fashion during preCollection. Each `Timer` object is also unique to each `AggregationProfileBreakdown` so there should be no thread safety issues there.

Reply:
@ticheng-aws would you mind double checking me on this here?

Reply:
@jed326, you can say that, and `buildAggregations()` in ProfilingAggregator.java uses the same `AggregationProfileBreakdown profileBreakdown` as `preCollection()`.